File: packages/sdk-py/src/uts_sdk/_types/status.py
# packages/sdk-py/src/uts_sdk/_types/status.py
"""Status enums and result types for timestamp verification."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import TYPE_CHECKING, Any

from .attestations import Attestation

if TYPE_CHECKING:
    from .timestamp_steps import Timestamp


class VerifyStatus(str, Enum):
    """Overall verification status for a timestamp."""

    VALID = "VALID"
    PARTIAL_VALID = "PARTIAL_VALID"
    INVALID = "INVALID"
    PENDING = "PENDING"


class UpgradeStatus(str, Enum):
    """Status of upgrading a pending attestation."""

    UPGRADED = "UPGRADED"
    PENDING = "PENDING"
    FAILED = "FAILED"


class AttestationStatusKind(str, Enum):
    """Status of a single attestation verification."""

    VALID = "VALID"
    INVALID = "INVALID"
    PENDING = "PENDING"
    UNKNOWN = "UNKNOWN"


class StampPhase(str, Enum):
    """Phase of a stamping operation."""

    QUEUED = "QUEUED"
    BATCHING = "BATCHING"
    AGGREGATING = "AGGREGATING"
    ATTESTING = "ATTESTING"
    CONFIRMING = "CONFIRMING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


class NodePosition(str, Enum):
    """Position of a node in a Merkle tree."""

    LEFT = "LEFT"
    RIGHT = "RIGHT"


@dataclass(frozen=True, slots=True, kw_only=True)
class AttestationStatus:
    """Result of attestation verification."""

    attestation: Attestation
    status: AttestationStatusKind
    error: Exception | None = None
    additional_info: dict[str, Any] | None = None


@dataclass(frozen=True, slots=True, kw_only=True)
class UpgradeResult:
    """Result of upgrading a pending attestation."""

    status: UpgradeStatus
    original: Attestation
    upgraded: Timestamp | None = None
    error: Exception | None = None

File: packages/sdk-go/types/status.go
package types

import "fmt"

type AttestationStatusKind int

const (
	StatusValid AttestationStatusKind = iota
	StatusInvalid
	StatusPending
	StatusUnknown
)

func (k AttestationStatusKind) String() string {
	switch k {
	case StatusValid:
		return "valid"
	case StatusInvalid:
		return "invalid"
	case StatusPending:
		return "pending"
	case StatusUnknown:
		return "unknown"
	default:
		return "unknown"
	}
}

type AttestationStatus struct {
	Attestation Attestation
	Status      AttestationStatusKind
	Error       error
	Info        map[string]interface{}
}

func NewAttestationStatus(att Attestation, status AttestationStatusKind, err error) *AttestationStatus {
	return &AttestationStatus{
		Attestation: att,
		Status:      status,
		Error:       err,
	}
}

func (s *AttestationStatus) String() string {
	if s.Error != nil {
		return fmt.Sprintf("%s: %s (%s)", s.Attestation, s.Status, s.Error)
	}
	return fmt.Sprintf("%s: %s", s.Attestation, s.Status)
}

type VerifyStatus int

const (
	VerifyValid VerifyStatus = iota
	VerifyPartialValid
	VerifyInvalid
	VerifyPending
)

func (s VerifyStatus) String() string {
	switch s {
	case VerifyValid:
		return "valid"
	case VerifyPartialValid:
		return "partial_valid"
	case VerifyInvalid:
		return "invalid"
	case VerifyPending:
		return "pending"
	default:
		return "unknown"
	}
}

type UpgradeStatus int

const (
	UpgradeUpgraded UpgradeStatus = iota
	UpgradePending
	UpgradeFailed
)

func (s UpgradeStatus) String() string {
	switch s {
	case UpgradeUpgraded:
		return "upgraded"
	case UpgradePending:
		return "pending"
	case UpgradeFailed:
		return "failed"
	default:
		return "unknown"
	}
}

type UpgradeResult struct {
	Status       UpgradeStatus
	Timestamp    *DetachedTimestamp
	Error        error
	Attestations []Attestation
}

func NewUpgradeResult(status UpgradeStatus, ts *DetachedTimestamp, err error) *UpgradeResult {
	return &UpgradeResult{
		Status:    status,
		Timestamp: ts,
		Error:     err,
	}
}

type VerificationResult struct {
	Status       VerifyStatus
	Attestations []*AttestationStatus
	Error        error
}

func NewVerificationResult(status VerifyStatus, atts []*AttestationStatus) *VerificationResult {
	return &VerificationResult{
		Status:       status,
		Attestations: atts,
	}
}

func (r *VerificationResult) HasValidAttestation() bool {
	for _, att := range r.Attestations {
		if att.Status == StatusValid {
			return true
		}
	}
	return false
}

func (r *VerificationResult) String() string {
	if r.Error != nil {
		return fmt.Sprintf("%s (error: %s)", r.Status, r.Error)
	}
	return fmt.Sprintf("%s (%d attestations)", r.Status, len(r.Attestations))
}
